### ISSUE TITLE
458. Poor Pigs - Solution and test cases are incorrect / don't match problem statement

### ISSUE THREAD

**Issue by jacobrshields** [LeetCode-Feedback/LeetCode-Feedback#8557]
#### Your LeetCode username

xxavalanchexx

#### Category of the bug
- [X] Question
- [X] Solution
- [ ] Language
- [ ] Missing Test Cases 

#### Description of the bug

The solution (and associated test cases) doesn't account for the fact that dead pigs cannot be reused in subsequent rounds.

According to the solution, `x` pigs with `s` available states can distinguish between `s^x` buckets.

Let `x` = 3 and `s` = 3. Then, allegedly, we should be able to distinguish between `3^3` = 27 buckets after 2 "rounds" of testing.

So we start with 27 buckets and 3 pigs. There are 8 unique outcomes for round 1: no pigs die; pig 1 dies; pig 2 dies; pig 3 dies; pigs 1 and 2 die; pigs 1 and 3 die; pigs 2 and 3 die; all pigs die. So, after round 1 the best we can do is narrow the poison down to `ceil(27/8)` = 4 buckets.

But if the outcome of round 1 was that all pigs died, now we have 3 dead pigs and we still need to search 4 buckets. Since we can't reuse dead pigs, we have to use more than 3 pigs total, so the solution is incorrect.

You may argue that the problem intended for the "best case scenario" instead. Well, if you're going for the best case scenario, the actual best case is that you assign 1 pig to 1 bucket at random, the pig dies after round 1, and you only needed 1 pig. So the "best case scenario" actually has a trivial solution.

In summary, the solution doesn't match the original problem statement.

**Comment by 152334H** [comment]
You're not supposed to assign the same number of buckets to each state. A state that is covered by `x-r` pigs will use up to `s**r` buckets.

Consider the breakdown for x=2, t=2:
1. Start with 9 buckets: [] [] [] [] [] [] [] [] []
2. If _all_ pigs die this round, then we cannot extract further information. So the state of [1,2] can only cover 1 bucket: [1,2] [] [] [] [] [] [] [] []
3. if _one_ pig dies this round, then we will only be able to extract info for 1 more bucket the next round. So the states of [1] and [2] can cover 2 buckets each: [1,2] [1] [1] [2] [2] [] [] [] []
4. Finally, if _no_ pigs die this round, the problem is reduced to x=2,t=1, where we know that info about **4** buckets can be extracted. So we leave the last 4 buckets unused on round 1.

**Comment by jacobrshields** [comment]
Ah I see, thank you!

So for my example of x=3, s=2, a.k.a. x=3, t=2, the assignments would be something like:

```
x=3 t=2

 1 [1,2,3]
 2 [1,2]
 3 [1,2]
 4 [1,3]
 5 [1,3]
 6 [2,3]
 7 [2,3]
 8 [1]
 9 [1]
10 [1]
11 [1]
12 [2]
13 [2]
14 [2]
15 [2]
16 [3]
17 [3]
18 [3]
19 [3]
20 []
21 []
22 []
23 []
24 []
25 []
26 []
27 []
```

**Comment by jacobrshields** [comment]
Closing this because I was wrong. Although I think the solution could be expanded to better explain how the pig->bucket assignments would actually work for multiple rounds, as explained by @152334H .
